### PR TITLE
Logs pagination support

### PR DESCRIPTION
**Pull request by PiotrDabkowski**
Full working version, but visual appearance has to be improved :)

Logs are now sent in chunks of 100 lines per request. User can move to older/never/newest/oldest logs by clicking appropriate buttons. 

Will add few more test cases.

---
This change is [Reviewable](https://reviewable.io/reviews/kubernetes/dashboard/1207)

### CODE REVIEW

**Comment by bryk** [comment]
Looks awesome! Will test this now :)

---

Reviewed 12 of 14 files at r1.
Review status: 12 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/app/backend/handler/apihandler.go, line 1189 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREH9-JYNw9Ds2Jlvai:-KREH9-KQkqV9qN5FY1K:bcrrtrp) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/handler/apihandler.go#L1189)):*
> ```Go
> 
> 	refTimestamp := request.QueryParameter("referenceTimestamp")
> 	refLineNum, _ := strconv.Atoi(request.QueryParameter("referenceLineNum"))
> ```

Don't ignore errors :) 

---

*[src/app/backend/resource/logs/logs.go, line 49 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREIWfJhQcYXiMB_ymh:-KREIWfJhQcYXiMB_ymi:b56ueu0) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L49)):*
> ```Go
> // MaxLogLines is a number that will be certainly bigger than any number of logs. Here 2 billion logs is certainly much larger
> // number of log lines than we can handle.
> var MaxLogLines int = 2000000000
> ```

Can you make this 2^52 -1? Or just a little less? 

---

*[src/app/backend/resource/logs/logs.go, line 60 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREIbTCjDqxGDTXOPcR:-KREIbTCjDqxGDTXOPcS:bob0b1x) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L60)):*
> ```Go
> 
> // Returns all logs.
> var NoLogViewSelector = &LogViewSelector{
> ```

Change "No" to "All"? 

---

*[src/app/backend/resource/logs/logs.go, line 92 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREIlMZDryTXyamGE1r:-KREIlMZDryTXyamGE1s:bnd2csi) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L92)):*
> ```Go
> 
> const (
> 	NewestTimestamp = ""
> ```

Why empty string? 

---

*[src/app/backend/resource/logs/logs.go, line 162 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREJvrPlU68SObQdu3p:-KREJvrQW9tpaen6hB5g:bd6659b) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L162)):*
> ```Go
> 		return matchingStartedAt + offset
> 	} else {
> 		return -1
> ```

Can you move -1 to a constant named, e.g., "LINE_INDEX_NOT_FOUND". And document this? Otherwise it is a magic number, hard to follow :) 

---

*[src/app/backend/resource/logs/logs.go, line 199 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREKW9xWNMea1iaEdKu:-KREKW9xWNMea1iaEdKv:b-737xh1) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L199)):*
> ```Go
> // SelectLogs returns selected part of LogLines as required by logSelector, moreover it returns IDs of first and last
> // of returned lines and the information of the resulting logView.
> func (self LogLines) SelectLogs(logSelector *LogViewSelector) (LogLines, LogLineId, LogLineId, LogViewInfo) {
> ```

Can you test this a bit? Corner and normal cases. 

---

*[src/app/backend/resource/logs/logs.go, line 218 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREKFC97hbtLbA2z-5P:-KREKFC97hbtLbA2z-5Q:b-3lqjsb) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L218)):*
> this part of array should not be affected by log deletion/addition.
 Why? Can you elaborate? 

And change 'should not' to 'cannot' if you're sure. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-:-KREKaBht-1-rZDIit0Y:b2mcqqy)*

**Comment by bryk** [comment]
Works awesome on serve:prod :)

**Comment by PiotrDabkowski** [comment]
@bryk  PTAL

---

Review status: 8 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/app/backend/handler/apihandler.go, line 1189 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREH9-JYNw9Ds2Jlvai:-KREWm1laYS2yq2KDhnJ:b-xkabme) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/handler/apihandler.go#L1189)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Don't ignore errors :) 
</details>
It will be 0 in case of error, this is what we want. But added lines that set it to 0 again in case of error anyway  :)

---

*[src/app/backend/resource/logs/logs.go, line 49 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREIWfJhQcYXiMB_ymh:-KREWx_RZEaQLiC8iH9B:bl511fv) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L49)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you make this 2^52 -1? Or just a little less? 
</details>
int value can be either int32 or int64, depending on platform, I can think of other solution if this is important.

---

*[src/app/backend/resource/logs/logs.go, line 60 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREIbTCjDqxGDTXOPcR:-KREXPHnfUmMrhztxgee:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L60)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Change "No" to "All"? 
</details>
Done.

---

*[src/app/backend/resource/logs/logs.go, line 92 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREIlMZDryTXyamGE1r:-KREXQOWG_AJkopRD8gZ:bnukzns) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L92)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Why empty string? 
</details>
NewestTimestamp is a default way of indexing, LogViewSelect without set RefLineId would use this method of indexing by default. I agree this might be a bit confusing so changed it  'newest'

---

*[src/app/backend/resource/logs/logs.go, line 162 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREJvrPlU68SObQdu3p:-KREXzirJ27X_fM81Bsc:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L162)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you move -1 to a constant named, e.g., "LINE_INDEX_NOT_FOUND". And document this? Otherwise it is a magic number, hard to follow :) 
</details>
Done.

---

*[src/app/backend/resource/logs/logs.go, line 199 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREKW9xWNMea1iaEdKu:-KREY-jw15nkqHTkovD4:bkq4gr5) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L199)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you test this a bit? Corner and normal cases. 
</details>
Added test cases to both frontend and backend.

---

*[src/app/backend/resource/logs/logs.go, line 218 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-KREKFC97hbtLbA2z-5P:-KREY2bz-wSgHfzasx57:b-mp3tr0) ([raw file](https://github.com/kubernetes/dashboard/blob/a07a45973fafcbe5f9b97ee79e8ef1261b0e48f0/src/app/backend/resource/logs/logs.go#L218)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> > this part of array should not be affected by log deletion/addition.
>  Why? Can you elaborate? 
> 
> And change 'should not' to 'cannot' if you're sure. 
</details>
We can never be sure :) Any line can be deleted, but the further away we are from the beginning the better :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-:-KREYFN8lBoT1mIGkp5Q:b-5u17mc)*

**Comment by bryk** [comment]
:lgtm:

---

Reviewed 1 of 14 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-:-KRSblDUzQlt3AH9dFUs:bnfp4nl)*

**Comment by bryk** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1207#-:-KRSfLsmwsWaz3k2J50A:bnfp4nl)*
